aplib.net-demo/Assets/Testing/AplibTests/SolveMazeTest.cs

```diff
@@ -0,0 +1,196 @@
+            }
+        );
+
+    /// <summary>
+    /// The name of the current room that the player is in.
+    /// </summary>
+    public Belief<Transform, string> CurrentRoom = new
+        (
+            reference: GameObject.Find("Player").transform,
+            getObservationFromReference: playerTransform =>
+            {
+                bool isHit = Physics.Raycast(playerTransform.position - new Vector3(0, Mathf.Epsilon, 0), -playerTransform.up, out RaycastHit hitRight, 4f);
+
+                if (isHit) return hitRight.collider.name;
+                else return "None";
+            }
+        );
+}
+
+public class SolveMazeTest
+{
+    [SetUp]
+    public void SetUp()
+    {
+        Debug.Log("Starting test SolveMazeTest");
+        SceneManager.LoadScene("SolveMazeTestScene");
+    }
+
+    /// <summary>
+    /// A UnityTest behaves like a coroutine in Play Mode. In Edit Mode you can use `yield return null;` to skip a frame.
+    /// </summary>
+    [UnityTest]
+    public IEnumerator PerformSolveMazeTest()
+    {
+        // Arrange
+        // Create a belief set for the agent.
+        SolveMazeBeliefSet beliefSet = new();
+
+        // Create an intent for the agent that moves the agent towards the target position.
+        Action<SolveMazeBeliefSet> rotateRightAction = new
+            (
+                effect: beliefSet =>
+                {
+                    Debug.Log("Rotating right");
+                    GameObject player = beliefSet.Player;
+                    player.transform.Rotate(0, 5, 0);
+                }
+            );
+        PrimitiveTactic<SolveMazeBeliefSet> rotateRight = new
+            (
+                action: rotateRightAction,
+                guard: beliefSet => !beliefSet.IsObstacleToRight
+            );
+
+        Action<SolveMazeBeliefSet> walkForwardAction = new
+            (
+                effect: beliefSet =>
+                {
+                    Debug.Log("Walking forward");
+                    GameObject player = beliefSet.Player;
+                    float playerSpeed = 10f;
+                    player.transform.position += playerSpeed * Time.deltaTime * player.transform.forward;
+                }
+            );
+        PrimitiveTactic<SolveMazeBeliefSet> walkForward = new
+            (
+                action: walkForwardAction,
+                guard: beliefSet => !beliefSet.IsObstacleInFront
+            );
+
+        Action<SolveMazeBeliefSet> rotateLeftAction = new
+            (
+                effect: beliefSet =>
+                {
+                    Debug.Log("Rotating left");
+                    GameObject player = beliefSet.Player;
+                    player.transform.Rotate(0, -5, 0);
+                }
+            );
+        PrimitiveTactic<SolveMazeBeliefSet> rotateLeft = new(action: rotateLeftAction);
+
+        FirstOfTactic<SolveMazeBeliefSet> walkAlongsideObstacles = new(null, rotateRight, walkForward, rotateLeft);
+
```

Comment:
Why the null? I would add `metadata: null'. The metadata is quite annoyting, and I feel tempted to propose to make it optional through an extension method. But that's not for this PR, and not a suggestion - just a rant